### ISSUE TITLE
Make nixos-rebuild ask for admin privileges

### ISSUE THREAD

**Issue by AMDphreak** [NixOS/nixpkgs#55802]
## Issue description

`nixos-rebuild` should ask for elevated privileges to execute.

> Additionally, it appears that you've used `nixos-rebuild` without `sudo` which is probably why you got the permission error.

citing this: **VSCode does not build (declarative installation)** #55790

**Comment by matthewbauer** [comment]
I think only nixos-rebuild switch needs root priveleges.

**Comment by vcunat** [comment]
Right, I commonly build my system without root privileges with `nixos-rebuild build`, and then I use `boot` or `switch` with `sudo`.  Adding `--upgrade` would also require `sudo`.

**Comment by AMDphreak** [comment]
> I think only nixos-rebuild switch needs root priveleges.

Good catch. I wasn't aware. I've been doing all commands as root.

**Comment by aaronduino** [comment]
Prompting the user for the sudo password at the beginning can save time. Think of how installing Windows would (at least used) take a while due to blocking dialogs showing up while the user is off doing another task (since the install takes a while).

On the other hand, the user may want to run as superuser only what's necessary, for security reasons. I  don't know if this is a legitimate concern once the entire the nixos-rebuild process is taken into account.

**Comment by aaronduino** [comment]
Relevant: https://discourse.nixos.org/t/error-on-nixos-rebuild-with-an-empty-configuration-file/1555

**Comment by AMDphreak** [comment]
@aaronjanse It does so much more. Requesting root is an essential user experience change. Linux land is filled with unhelpful error messages. This habit has to be fixed.

**Comment by domenkozar** [comment]
Relevant: https://github.com/hercules-ci/ghcide-nix/issues/14#issuecomment-542912692

**Comment by bobvanderlinden** [comment]
It seems there is [a conscious decision to not automatically run sudo](https://github.com/NixOS/nixpkgs/pull/57333#issuecomment-493048194), but make it explicit.

It is possible to run sudo automatically when using `--use-remote-sudo`:

```
$ nixos-rebuild --use-remote-sudo boot
```

**Comment by con-f-use** [comment]
The problem with that `nixos-rebuild --use-remote-sudo` is broken (see https://github.com/NixOS/nixpkgs/issues/118655)

**Comment by rgri** [comment]
Curious if there are any favourite workarounds people are using for this. I suppose I will just end up using a [deployment tool](https://github.com/nix-community/awesome-nix#deployment-tools), but my scale is reallly small, so if I could deploy using just `nixos-rebuild` I'd be very happy with that.

**Comment by loic-roux-404** [comment]
bump 

It could be so nice to provide root password to let nix make the escalation during rebuild giving password to prompt automaticly like in ansible
